File: NVP/src/Uebung9/philSTM.hs
import Control.Concurrent
import Control.Concurrent.STM

type Stick = TVar Bool

main :: IO ()
main = do
  s1 <- atomically newStick
  s2 <- atomically newStick
  s3 <- atomically newStick
  s4 <- atomically newStick
  s5 <- atomically newStick
  forkIO (phil s1 s2 1)
  forkIO (phil s2 s3 2)
  forkIO (phil s3 s4 3)
  forkIO (phil s4 s5 4)
  getLine
  phil s5 s1 5
  
newStick :: STM Stick
newStick = do
  newTVar True

takeStick stick = do
  available <- readTVar stick
  if available then do
    writeTVar stick False
  else do
    retry

putStick :: Stick -> STM ()
putStick stick = do
  writeTVar stick True

phil :: Stick -> Stick -> Int -> IO ()
phil sl sr nr = do
  putStrLn (show nr ++ " is thinking")
  atomically $ do takeStick sl
                  takeStick sr
  --atomically $ takeStick sl
  --atomically $ takeStick sr
  putStrLn (show nr ++ " is eating")
  threadDelay 1000
  atomically $ putStick sl
  atomically $ putStick sr
  phil sl sr nr

File: NVP/src/Uebung10/aufgabe1/philSTM.erl
-module(philSTM).
-export([new_stick/0,take/1,put/1,start_phils/1]).
-import(stm,[new_tvar/1,read_tvar/1,write_tvar/2,retry/0,atomically/1]).

take(S) ->
  V = read_tvar(S),
  case V of
    true -> write_tvar(S,false);
    false -> retry()
  end.

put(S) ->
  write_tvar(S,true).

new_stick() -> atomically(fun() -> new_tvar(true) end).

phil(SL,SR,N) -> 
  base:printLn(base:show(N)++" is thinking."),
  atomically(fun() -> take(SL) end),
  atomically(fun() -> take(SR)
             end),
  base:printLn(base:show(N)++" is eating."),
  atomically(fun() -> put(SR),
                      put(SL)
             end),
  phil(SL,SR,N).

start_phils(N) -> 
  Sticks = create_sticks(N),
  Last_stick = create_phils(Sticks,2),
  base:getLn("Start last phil >"),
  phil(Last_stick,hd(Sticks),1).

create_sticks(0) -> [];
create_sticks(N) -> [new_stick() | create_sticks(N-1)].

create_phils([S1,S2|Sticks],N) ->
  spawn(fun() -> phil(S1,S2,N) end),
  create_phils([S2|Sticks],N+1);
create_phils([Last_stick],_) -> Last_stick.

















